src/js/libcs/Json.js

```diff
@@ -0,0 +1,223 @@
+// CindyScript JSON
+var Json = {};
+Json._helper = {};
+
+
+Json.turnIntoCSJson = function(a) {
+    return {
+        "ctype": "JSON",
+        "value": a
+    };
+};
+
+Json._helper.ShallowClone = function(o) {
+    var out, v, key;
+    out = Array.isArray(o) ? [] : {};
+    for (key in o) {
+        v = o[key];
+        out[key] = v;
+    }
+    return out;
+};
+
+Json._helper.DeepClone = function(o) {
+    var out, v, key;
+    out = Array.isArray(o) ? [] : {};
+    for (key in o) {
+        v = o[key];
+        out[key] = (typeof v === "object" && v !== null) ? Json._helper.DeepClone(v) : v;
+    }
+    return out;
+};
+
+Json.getField = function(obj, key) {
+    if (obj.value && obj.value[key]) {
+        return obj.value[key];
+    }
+    return nada;
+};
+
+Json.setField = function(where, field, what) {
+    where[field] = what;
+};
+
+Json.GenFromUserDataEl = function(el) {
+    // key/obj are reversed due to the semantics of the ":" operator in CindyScript
+    var key = el.obj;
+    var obj = el.key;
+
+    if (!key || key.ctype !== "string") {
+        console.log("Error: JSON keys have to be strings.");
+        return nada;
+    }
+    if (!obj) {
+        console.log("Warning: JSON object not defined.");
+        return {
+            "key": key.value,
+            "val": nada,
+        };
+    } else return {
+        "key": key.value,
+        "val": evaluate(obj)
+    };
+};
+
+Json._helper.GenJSONAtom = function(key, val) {
+    return {
+        "ctype": "JSONAtom",
+        "value": {
+            'key': key,
+            'value': val
+        }
+    };
+};
+
+
+Json._helper.forall = function(li, runVar, fct, modifs) { // JSON
+    // default iterate over values in JSON
+    var iteratorType = "value";
+    var res;
+    if (modifs.iterator !== undefined) {
+        let it = evaluate(modifs.iterator);
+        let iterTypes = ["key", "value", "pair"];
+        if (it.ctype === 'string' && iterTypes.includes(it.value)) {
+            iteratorType = it.value;
+        }
+    }
+    // iterate over key, value or pair
+    if (iteratorType === "value") {
+        for (let k in li) {
+            namespace.setvar(runVar, li[k]);
+            res = evaluate(fct);
+        }
+    } else if (iteratorType === "key") {
+        for (let k in li) {
+            namespace.setvar(runVar, General.string(k));
+            res = evaluate(fct);
+        }
+    } else { // pair
+        for (let k in li) {
+            namespace.setvar(runVar, Json._helper.GenJSONAtom(k, li[k]));
```

Comment:
Do we really need JSONAtoms? Because having them, a lot of things would be needed to changed as well. For instance, the code

```
a = {"a":2, "b":3};
forall(a, x, iterator->"pair",
      x.value=x.value^2;
      errc(x.value)
);
``` does give currently `2` and `3`. However, I would expect `4` and `9`.

Would not it be easier if these pair-iterators are just plain JSON-objects them-self with having the keys `key` and `value`? (the only disadvantage I see is that niceprint of "just-JSON-iterators" would look like `{key:a, value:2}`, which is okay IMO because it would help developers to debug.)